### ISSUE TITLE
"var let" should be allowed in non-strict mode [$15]

### ISSUE THREAD

**Issue by DanielRosenwasser** [eslint/espree#149]
Repro:

```JavaScript
var let = 10;
```

See https://github.com/jquery/esprima/issues/1048

**Comment by nzakas** [comment]
I believe this happens because `let` token is marked as a keyword, so this check fails:

https://github.com/eslint/espree/blob/master/espree.js#L3596

I think this might be as simple as removing "let" from this statement:

https://github.com/eslint/espree/blob/master/lib/syntax.js#L154

@btmills @xjamundx @FredKSchott do any of you have time to look at fixing this?

**Comment by xjamundx** [comment]
Early next week if no one claims it first.

**Comment by xjamundx** [comment]
Assuming `var const = 10` should also work?

**Comment by DanielRosenwasser** [comment]
@xjamundx no, `const` is a future reserved word. I can send out a PR. My hesitation in doing so was really due to the issue that this behavior was originally supposed to be for compatibility with the SpiderMonkey AST API.

**Comment by DanielRosenwasser** [comment]
Actually, there are some implications to that original suggestion that make this a little more involved:

* The tokenizer now no longer categorizes `let` as a keyword even though this may be ideal for most purposes.
* More places must be checked for `let` as an identifier. See below.

I'll do some more work on this, but given that I'm largely unfamiliar with the code, I might end up handing this back off to you :smile: 

-----------------
-----------------

> ### <span id="sec-13.2.1.1" class="secnum">[13.2.1.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-let-and-const-declarations-static-semantics-early-errors "link to this section")</span> Static Semantics: Early Errors

> *LexicalDeclaration* <span class="geq">:</span> *LetOrConst* *BindingList* `;`

> -   It is a Syntax Error if the BoundNames of *BindingList* contains `"let"`.

-----------------

> ### <span id="sec-13.6.4.1" class="secnum">[13.6.4.1](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-for-in-and-for-of-statements-static-semantics-early-errors "link to this section")</span> Static Semantics: Early Errors

> *IterationStatement* <span class="geq">:</span>

> `for` `(` *ForDeclaration* `in` *Expression* `)` *Statement*

> `for` `(` *ForDeclaration* `of` *AssignmentExpression* `)` *Statement*

> -   It is a Syntax Error if the BoundNames of *ForDeclaration* contains `"let"`.

**Comment by nzakas** [comment]
OK, so probably makes sense to keep `let` as a keyword and just add an exception that looks for an identifier or `let` in `var` declarations outside of loops.

**Comment by xjamundx** [comment]
And by "early next week"....I'll try again in the coming week.

**Comment by nzakas** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
The thing is that something like `let + 100` is valid, so it's not always a keyword.

**Comment by nzakas** [comment]
I wonder if Acorn handles this correctly. Might be useful to see what they're doing.

**Comment by xjamundx** [comment]
Cool. Will check on how they're doing it!

**Comment by xjamundx** [comment]
With the latest version of acorn `2.1.1`:

```
> require('./dist/acorn').parse('var let = y', {ecmaVersion: '6'})
SyntaxError: Unexpected token (1:4)
    at Parser.pp.raise (/Users/jamuferguson/dev/acorn/dist/acorn.js:975:13)
    at Parser.pp.unexpected (/Users/jamuferguson/dev/acorn/dist/acorn.js:1524:8)
    at Parser.pp.parseBindingAtom (/Users/jamuferguson/dev/acorn/dist/acorn.js:1164:12)
    at Parser.pp.parseVarId (/Users/jamuferguson/dev/acorn/dist/acorn.js:2041:18)
    at Parser.pp.parseVar (/Users/jamuferguson/dev/acorn/dist/acorn.js:2024:10)
    at Parser.pp.parseVarStatement (/Users/jamuferguson/dev/acorn/dist/acorn.js:1913:8)
    at Parser.pp.parseStatement (/Users/jamuferguson/dev/acorn/dist/acorn.js:1713:19)
    at Parser.pp.parseTopLevel (/Users/jamuferguson/dev/acorn/dist/acorn.js:1653:21)
    at Parser.parse (/Users/jamuferguson/dev/acorn/dist/acorn.js:1623:17)
    at Object.parse (/Users/jamuferguson/dev/acorn/dist/acorn.js:937:44)
```

**Comment by xjamundx** [comment]
I sort of gave up this when thinking harder about the complexity. May take another stab if I get motivated today though after looking at some of the other espree isues...

**Comment by xjamundx** [comment]
Going to throw up a PR with broken tests, just to keep me motivated....

**Comment by nzakas** [comment]
Acorn issue: https://github.com/ternjs/acorn/issues/359

**Comment by DanielRosenwasser** [comment]
As a heads up, this was fixed in esprima.

**Comment by michaelficarra** [comment]
I just got bit by this with

```js
(let[a] = b);
```

This should be valid, but is giving me the error "Parsing error: Unexpected token let".

**Comment by nzakas** [comment]
Yup, it's an upstream issue with Acorn.

**Comment by RReverser** [comment]
IIUC, this issue can be closed now.

**Comment by mysticatea** [comment]
I confirmed it.

```js
"use strict";

const espree = require("espree");
const ast = espree.parse(
    `var let = 0`,
    {
        ecmaVersion: 6,
        sourceType: "script"
    }
);

console.log(JSON.stringify(ast, null, 4));
```

```json
{
    "type": "Program",
    "start": 0,
    "end": 11,
    "body": [
        {
            "type": "VariableDeclaration",
            "start": 0,
            "end": 11,
            "declarations": [
                {
                    "type": "VariableDeclarator",
                    "start": 4,
                    "end": 11,
                    "id": {
                        "type": "Identifier",
                        "start": 4,
                        "end": 7,
                        "name": "let"
                    },
                    "init": {
                        "type": "Literal",
                        "start": 10,
                        "end": 11,
                        "value": 0,
                        "raw": "0"
                    }
                }
            ],
            "kind": "var"
        }
    ],
    "sourceType": "script"
}
```

**Comment by michaelficarra** [comment]
@mysticatea how about the program I mentioned?

**Comment by RReverser** [comment]
@michaelficarra It works http://astexplorer.net/#/5L9D8euj22. Been fixed in https://github.com/ternjs/acorn/commit/8f46eb5ffa5248a9b743b7e8f3e464620ce37c5c couple of months ago.

**Comment by nzakas** [comment]
I'll add a test case just to be sure.

**Comment by RReverser** [comment]
I guess money goes to @marijnh in such case :smile:

**Comment by nzakas** [comment]
Yup. @marijnh, once this is closed, you can claim the bounty on this issue.
